### PR TITLE
Initial Commit

### PR DESCRIPTION
**Pull request by captain-pool**
Initial Commit
----------------------
Evaluation Phase: 1
Contents:
- Added Exporter for ONNX Model for ShuffleNet
- Added Sample TF Hub Exporter for TF 2.0
- Added Sample for Image Retraining using TF Hub, Tensorflow 2.0 and TPUEstimator

### CODE REVIEW

File: Evaluation_1/ShuffleNet/export.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+  tar.extractall()
+  tar.close()
+
+
+if not os.path.exists("shufflenet.pb"):
+  model = onnx.load("shufflenet/model.onnx")
+  tf_rep = prepare(model)
+  tf_rep.export_graph("shufflenet.pb")
+
+
```
Comment (srjoglekar246) [inline Evaluation_1/ShuffleNet/export.py]: For readability, try to restructure files as:
1] All function definitions & helpers
2] An `if __name__ == "__main__"` block (that defines what the binary does).

Also, try to have a max of 2 newlines between two functions/blocks.

File: Evaluation_1/TPU_Samples/README.md [hunk 2]
```diff
@@ -0,0 +1,2 @@
+# Image Retraining with Tensorflow 2.0, TF Hub and Cloud TPU
```
Comment (srjoglekar246) [inline Evaluation_1/TPU_Samples/README.md]: Also add instructions in these READMEs about stuff like high-level requirements (TPU access on Cloud, etc), commands to run, etc. This will make these scripts way more useful.
This can be done in a followup PR, just add a line here like:
```
TODO(captain-pool@): <...>
```

File: Evaluation_1/TPU_Samples/image_retraining_tpu.py [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+flags.DEFINE_integer("iterations",2, "Number of Itertions")
+flags.DEFINE_integer("batch_size", 10, "Size of eahc Batch")
+flags.DEFINE_boolean("use_tpu", True, " Use TPU")
+flags.DEFINE_string("model_dir", "model_dir/", "Directory to Save the Models and Checkpoint")
+flags.DEFINE_string("dataset",
+            "horses_or_humans",
+            "TFDS Dataset Name. IMAGE Dimension should be >= 224, channel=3")
+flags.DEFINE_string("data_dir", None, "Directory to Save Data to")
+NUM_CLASSES = None
+FEATURE_SHAPE = None
```
Comment (srjoglekar246) [inline Evaluation_1/TPU_Samples/image_retraining_tpu.py]: Newlines between flags, constants & function definitions.

File: Evaluation_1/TPU_Samples/image_retraining_tpu.py [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+          input_fn=lambda params:input_fn(
+              mode=tf.estimator.ModeKeys.TRAIN,
+              **params),
+          max_steps=1000)#.evaluate(
+                  #input_fn=lambda params:input_fn(
+                  #    mode=tf.estimator.ModeKeys.EVAL,
+                  #    **params),
+                  #steps=1000)
+  classifier.export_savedmodel(os.path.join(FLAGS.model_dir,"export"), serving_reciever_input_fn)
+if __name__ == "__main__":
```
Comment (srjoglekar246) [inline Evaluation_1/TPU_Samples/image_retraining_tpu.py]: Newlines here.

File: Evaluation_1/TFHub_Sample_Deploy/export.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
+import tensorflow as tf
+import tensorflow_datasets as tfds
+import tensorflow_hub as hub
+""" Create a Sample TF-Hub Module using SavedModel v2.0
+The module has as a single signature which loads MNIST Dataset from TFDS and train a simple Neural Network for classifying the digits. The model is built and trained using Tewnsorlfow 
+"""
+
+class MNIST(tf.keras.models.Model):
```
Comment (srjoglekar246) [inline Evaluation_1/TFHub_Sample_Deploy/export.py]: Add doctrings to classes & functions. If a lot is in flux, file an Issue on the repo to do this later.

File: Evaluation_1/TPU_Samples/image_retraining_tpu.py [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+      "dataset": FLAGS.dataset
+    }
+  )
+
+  classifier.train(
+          input_fn=lambda params:input_fn(
+              mode=tf.estimator.ModeKeys.TRAIN,
+              **params),
+          max_steps=1000)#.evaluate(
+                  #input_fn=lambda params:input_fn(
```
Comment (srjoglekar246) [inline Evaluation_1/TPU_Samples/image_retraining_tpu.py]: Remove commented-out code?
If its something that will be enabled later, add a TODO (see other comment)

**Comment by srjoglekar246** [comment]
Can you improve the description of the PR? What things are being added, etc.
Also, optional: it would be useful to have Issues for all your projects, so all our discussion etc is tracked and archived :-). Chats and emails are worse to find stuff in.

File: Evaluation_1/TFHub_Sample_Deploy/export.py (hunk 5 above)
Comment (captain-pool) [reply Evaluation_1/TFHub_Sample_Deploy/export.py]: Created an Issue for it :)

File: Evaluation_1/TPU_Samples/image_retraining_tpu.py (hunk 6 above)
Comment (captain-pool) [reply Evaluation_1/TPU_Samples/image_retraining_tpu.py]: Done

File: Evaluation_1/TPU_Samples/image_retraining_tpu.py (hunk 4 above)
Comment (captain-pool) [reply Evaluation_1/TPU_Samples/image_retraining_tpu.py]: Done

File: E1_ShuffleNet/export.py [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+import os
+import io
+import requests
+from tqdm import tqdm
+import tarfile
+
+DOWNLOAD_LINK = "https://s3.amazonaws.com/download.onnx/models/opset_8/shufflenet.tar.gz"
+
+
+if not os.path.exists(DOWNLOAD_LINK.split("/")[-1]):
```
Comment (srjoglekar246) [inline E1_ShuffleNet/export.py]: This kind of code can make debugging tricky, since its essentially doing the job of 'main' without being included under the `main()` function. For example, in these blocks you download ShuffleNet into a protobuf file `shufflenet.pb`, which is assumed to exist in another helper.

Instead, have a function as `load_shufflenet()` that downloads (or loads) the protobuf file, and make `shufflenet.pb` a constant `SHUFFLENET_PB` thats referenced everywhere.

File: E1_TFHub_Sample_Deploy/export.py [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+  metric(loss_)
+
+
+@tf.function
+def test(image, label):
+  preds = model(image)
+  label_onehot = tf.one_hot(label, 10)
+
+
+for epoch in range(10):
```
Comment (srjoglekar246) [inline E1_TFHub_Sample_Deploy/export.py]: Same comment as the other file. Try to have all your code in either:
1. A helper
OR
2. A `main()` function.

I know its quicker to write code this way, but these conventions make things easier for your future self and others who will use this code :-)

File: E1_TPU_Samples/README.md [hunk 9]
```diff
@@ -0,0 +1,84 @@
+# Image Retraining Sample
```
Comment (srjoglekar246) [inline E1_TPU_Samples/README.md]: Awesome work on these READMEs!

File: E1_TPU_Samples/image_retraining_tpu.py [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+  model = tf.keras.Sequential([
+      hub.KerasLayer("https://tfhub.dev/google/tf2-preview/inception_v3/feature_vector/4",
+                     output_shape=[2048],
+                     trainable=False
+                     ),
+      tf.keras.layers.Dense(NUM_CLASSES, activation="softmax")
+  ])
+  optimizer = None
+  if mode == tf.estimator.ModeKeys.TRAIN:
+    optimizer = tf.optimizers.Adam(params.get("learing_rate", 1e-3))
```
Comment (srjoglekar246) [inline E1_TPU_Samples/image_retraining_tpu.py]: Does this work without your hack in TF? Sorry, I can't remember what we landed on.
Maybe add a TODO mentioning the Issue you filed, and say you will refactor once it is fixed. (And if possible, use compat functionality until the Issue is fixed, so someone could just pull these scripts and run them as-is).

**Comment by srjoglekar246** [comment]
(Also CC Vojtech on this PR, to keep him in the loop)

File: E1_TPU_Samples/README.md (hunk 9 above)
Comment (captain-pool) [reply E1_TPU_Samples/README.md]: :)

File: E1_TPU_Samples/image_retraining_tpu.py (hunk 10 above)
Comment (captain-pool) [reply E1_TPU_Samples/image_retraining_tpu.py]: This doesn't really work without the Hack :(
I can use compat module for this.

**Comment by captain-pool** [comment]
> (Also CC Vojtech on this PR, to keep him in the loop)

Vojtech has not accepted the collaborator's request yet. He might be out of office. He have not yet answered to hangouts chat.

**Comment by captain-pool** [comment]
> Good job with the code refactoring. Remaining comments are mainly about getting your other code to look like `image_retraining_tpu.py` :-)

Bash to rescue really. haha!

For Refactoring and Linting I'm using:
```bash
for i in $(find . -maxdepth 2 -name "*.py"); do autopep8 -i -a -a --indent-size=2 $i; done;
```

And for that Directory Structure Updating, I used:
```bash
for i in $(find . -maxdepth 1 -type d); do i=$(basename $i); mv $i ../E1_$i; done;
```
and removed that `Evaluation_1` Directory :smile:

File: E1_TPU_Samples/image_retraining_tpu.py (hunk 10 above)
Comment (captain-pool) [reply E1_TPU_Samples/image_retraining_tpu.py]: I added a command line flag `--use_compat`  to switch from hack mode to compat mode. I think that would help. :)

File: E1_ShuffleNet/export.py (hunk 7 above)
Comment (captain-pool) [reply E1_ShuffleNet/export.py]: Done :)
